javalib/src/main/scala/java/io/DataInputStream.scala

```diff
@@ -35,14 +35,21 @@ class DataInputStream(in: InputStream)
     readFully(b, 0, b.length)
 
   override final def readFully(b: Array[Byte], off: Int, len: Int): Unit = {
-    if (off < 0 || len < 0 || len > b.length - off)
+    if ((off < 0) || (len < 0) || (len > b.length - off)) {
       throw new IndexOutOfBoundsException()
-    var i = 0
-    while (i < len) {
-      val value = readByte()
-      if (value == -1) throw new EOFException()
-      b(i + off) = value
-      i += 1
+    }
+
+    var offset = off
+
+    while (offset < len) {
```

Comment:
I don't think this is correct. `off` argument gives and offset potentially in the middle of the array, `len` is not necessarily the length of the array but number of the bytes to read. So for example in an array with 20 elements, you can give offset 10 and length 2. This would mean that the while loop is not going to run even a single iteration.

Reply:
A better version of the same code would be:
```
  override final def readFully(b: Array[Byte], off: Int, len: Int): Unit = {
    if ((off < 0) || (len < 0) || (len > b.length - off)) {
      throw new IndexOutOfBoundsException()
    }

    var offset = off
    var length = len

    while (length > 0) {
      val read = in.read(b, offset, length)

      if (read == -1) {
        throw new EOFException()
      } else {
        offset += read
        length -= read
      }
    }
  }
```

Reply:
You are absolutely correct! Space shot on my part, sorry.

I will make the suggested correction this evening (Eastern Daylight Time).

Reply:
Thank you for the correction.